File: app/src/main/java/com/futsch1/medtimer/helpers/IdlingListAdapter.java
package com.futsch1.medtimer.helpers;

import androidx.annotation.NonNull;
import androidx.annotation.Nullable;
import androidx.recyclerview.widget.DiffUtil;
import androidx.recyclerview.widget.ListAdapter;
import androidx.recyclerview.widget.RecyclerView;

import java.util.List;

@SuppressWarnings("java:S119") // Reproducing the naming of the ListAdapter class
public abstract class IdlingListAdapter<T, VH extends RecyclerView.ViewHolder>
        extends ListAdapter<T, VH> {

    private final SimpleIdlingResource idlingResource;

    protected IdlingListAdapter(@NonNull DiffUtil.ItemCallback<T> diffCallback) {
        this(diffCallback, "IdlingListAdapter_" + diffCallback);
    }

    protected IdlingListAdapter(@NonNull DiffUtil.ItemCallback<T> diffCallback, String idlingResourceName) {
        super(diffCallback);
        idlingResource = IdlingResourcesPool.getInstance().getResource(idlingResourceName);
        idlingResource.setBusy();
    }

    @Override
    public void submitList(@Nullable List<T> list) {
        super.submitList(list);
        idlingResource.setIdle();
    }

    @Override
    public void submitList(@Nullable List<T> list, Runnable commitCallback) {
        super.submitList(list, commitCallback);
        idlingResource.setIdle();
    }

    public void resetInitialized() {
        idlingResource.setBusy();
    }
}

File: app/src/main/java/com/futsch1/medtimer/helpers/IdlingListAdapter.kt
package com.futsch1.medtimer.helpers

import androidx.recyclerview.widget.DiffUtil
import androidx.recyclerview.widget.ListAdapter
import androidx.recyclerview.widget.RecyclerView
import com.futsch1.medtimer.helpers.IdlingResourcesPool.Companion.getInstance

abstract class IdlingListAdapter<T, VH : RecyclerView.ViewHolder?>
protected constructor(diffCallback: DiffUtil.ItemCallback<T?>, idlingResourceName: String = "IdlingListAdapter_$diffCallback") :
    ListAdapter<T?, VH?>(diffCallback) {
    private val idlingResource: SimpleIdlingResource = getInstance().getResource(idlingResourceName)

    init {
        idlingResource.setBusy()
    }

    override fun submitList(list: MutableList<T?>?) {
        super.submitList(list)
        idlingResource.setIdle()
    }

    override fun submitList(list: MutableList<T?>?, commitCallback: Runnable?) {
        super.submitList(list, commitCallback)
        idlingResource.setIdle()
    }
}